packages/Php70/src/Rector/FuncCall/NonVariableToVariableOnFunctionCallRector.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Php70\Rector\FuncCall;
+
+use function lcfirst;
+use PhpParser\Node;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\ArrayDimFetch;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\AssignOp;
+use PhpParser\Node\Expr\AssignRef;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Expr\PropertyFetch;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Name;
+use PhpParser\Node\Stmt\Return_;
+use PHPStan\Analyser\MutatingScope;
+use PHPStan\Analyser\Scope;
+use PHPStan\Broker\Broker;
+use PHPStan\Broker\FunctionNotFoundException;
+use PHPStan\Reflection\ParameterReflection;
+use PHPStan\Reflection\ParametersAcceptor;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use Rector\CodingStyle\Naming\ClassNaming;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Rector\Php70\Dto\VariableAssignPair;
+use Rector\Rector\AbstractRector;
+use Rector\RectorDefinition\CodeSample;
+use Rector\RectorDefinition\RectorDefinition;
+
+/**
+ * @see https://www.php.net/manual/en/migration70.incompatible.php
+ *
+ * @see \Rector\Php70\Tests\Rector\FuncCall\NonVariableToVariableOnFunctionCallRector\NonVariableToVariableOnFunctionCallRectorTest
+ */
+final class NonVariableToVariableOnFunctionCallRector extends AbstractRector
+{
+    /**
+     * @var string
+     */
+    private const DEFAULT_VARIABLE_NAME = 'tmp';
+
+    /**
+     * @var Broker
+     */
+    private $broker;
+
+    /**
+     * @var ClassNaming
+     */
+    private $classNaming;
+
+    public function __construct(Broker $broker, ClassNaming $classNaming)
+    {
+        $this->broker = $broker;
+        $this->classNaming = $classNaming;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition(
+            'Transform non variable like arguments to variable where a function or method expects an argument passed by reference',
+            [new CodeSample('reset(a());', '$a = a(); reset($a);')]
+        );
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [FuncCall::class, MethodCall::class, StaticCall::class];
+    }
+
+    /**
+     * @param FuncCall|MethodCall|StaticCall $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = $node->getAttribute(AttributeKey::SCOPE);
+        if (! $scope instanceof MutatingScope) {
```

Comment:
`Scope` is good enough as interface

Reply:
`Scope` doesn't contains the method `assignExpression` sadly. See (upcoming) explanation below.

Reply:
Ah, I see. I'm affraid that this will go in mess like "which one to use when", since `instanceof Scope` is used everywhere now.

Reply:
Will see if I can do something with the `ScopeFactory` instead.

Reply:
But a simple rule could be if you need to mutate scope, go with `MutatingScope`, otherwise use `Scope`.

Reply:
It might, but classes with more public method than interfaces they implement always create crappy code. It's PHPStan design flaw we cannot cover here I'm affraid.

Let's keep it your way for now, not that important.

Reply:
Furthermore, `MutatingScope` seems to be the only available implementation of `Scope`.